### PR TITLE
AuthPolicy: add validation and allow multiple actions per policy

### PR DESCRIPTION
**Pull request by rahulanand16nov**
Small hardening for #133 

User cannot target rejected HTTPRoute and hostname from the route is passed to IAP (Istio's AuthPolicy) now.

## Verification Steps
1. `make local-setup`
2. `k apply -f example/toystore/toystore.yaml`
3. `k apply -f example/toystore/httproute.yaml`
4. `k apply -f config/samples/apim_v1alpha1_authpolicy.yaml`
5. `k apply -f config/samples/apim_v1alpha1_authpolicy.yaml`

In the 5th step, you'll see the hostname in IAP but it's not present in the Kuadrant's AuthPolicy hence taken from HTTPRoute.

### CODE REVIEW

File: controllers/apim/utils.go [hunk 1]
```diff
@@ -48,3 +50,15 @@ func alwaysUpdateRateLimit(existingObj, desiredObj client.Object) (bool, error)
...
 	return true, nil
 }
+
+func TargetableRoute(httpRoute *gatewayapiv1alpha2.HTTPRoute) error {
+	for _, parent := range httpRoute.Status.Parents { // no parent mean policies will affect nothing.
+		if len(parent.Conditions) == 0 {
+			return fmt.Errorf("unable to verify targetability: no condition found on status")
+		}
+		if meta.IsStatusConditionFalse(parent.Conditions, "Accepted") {
+			return fmt.Errorf("httproute rejected")
```
Comment (eguzki) [inline controllers/apim/utils.go:60]: nitpìck: I thought that `fmt.Errorf` required at least one placeholder. If no placeholders are needed, then `errros.New` is the error builder.

File: controllers/apim/authpolicy_controller.go [hunk 2]
```diff
@@ -105,6 +119,12 @@ func (r *AuthPolicyReconciler) reconcileAuthPolicy(ctx context.Context, ap *apim
 		// convert []Rule  to []*Rule
 		rulePtrSlice := []*secv1beta1types.Rule{}
 		for idx := range ap.Spec.Rules {
+			// Force hostname usage from HTTPRoute
+			for _, toRule := range ap.Spec.Rules[idx].To {
```
[the remark this answers is not in the document]
Comment (rahulanand16nov) [reply controllers/apim/authpolicy_controller.go:123]: I agree the current behavior is wrong. We should check the regex and then append it. Added a note for now.

File: controllers/apim/authpolicy_controller.go [hunk 3]
```diff
@@ -137,6 +157,39 @@ func (r *AuthPolicyReconciler) reconcileAuthPolicy(ctx context.Context, ap *apim
...
+	httpRouteAnnotations := httpRoute.GetAnnotations()
+	if httpRouteAnnotations == nil {
+		httpRouteAnnotations = map[string]string{}
+	}
+
+	apKey := client.ObjectKeyFromObject(ap)
+	val, present := httpRouteAnnotations[common.AuthPolicyBackRefAnnotation]
+	if present {
+		if val != apKey.String() {
+			return fmt.Errorf("the target HTTPRoute {%s} is already referenced by authpolicy %s", client.ObjectKeyFromObject(httpRoute), apKey.String())
```
[the remark this answers is not in the document]
Comment (rahulanand16nov) [reply controllers/apim/authpolicy_controller.go:186]: I want to align authpolicy with the concept described in the policy attachment as much as possible i.e. one resource per policy. Made changes that allow multiple actions per resource now.

File: controllers/apim/utils.go (hunk 1 above)
Comment (rahulanand16nov) [reply controllers/apim/utils.go]: Fixed

**Review by eguzki — APPROVED** [review]

**Pull request merged**